After upgrading jib-maven-plugin from 3.0.0 to 3.1.0 entrypoint=INHERIT doesn't ignore mainClass

<!--
Before filing an issue, please reproduce with the latest version of Jib.

If you are encountering errors pulling or pushing to remote registries,
please check our Frequently Asked Questions before filing an issue:

    https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#registry-errors
-->

**Environment**:

- *Jib version:* 3.1.0 (jib-maven-plugin)
- *Build tool:* Maven 3.6.3
- *OS:* Linux

**Description of the issue**:

After updating `jib-maven-plugin` from version `3.0.0` to version `3.1.0` execution of `dockerBuild` goal for container with container.entrypoint=INHERIT configuration option fails with error:

```
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:3.1.0:dockerBuild (docker-image-package) on project javadocs: Main class was not found, perhaps you should add a `mainClass` configuration to jib-maven-plugin -> [Help 1]
```

[README says](https://github.com/GoogleContainerTools/jib/blob/master/jib-maven-plugin/README.md#container-object) that entrypoint=INHERIT means mainClass is not taken into account, which seems to be broken in 3.1.0

**Expected behavior**:

After updating 3.0.0 -> 3.1.0 build should succeed because release notes don't say anything about changed behaviour of entrypoint=INHERIT configuration.

**Steps to reproduce**:
<!-- Please provide a minimal and precise series of steps -->

1. Add `jib-maven-plugin` configuration from block below with version `3.0.0` into pom.xml:

2. Execute the following command: `mvn package` and Docker image is **built successfully**.

3. Change `jib-maven-plugin` version to `3.1.0`, execute `mvn package` - **build fails** with the following error:
```
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:3.1.0:dockerBuild (docker-image-package) on project javadocs: Main class was not found, perhaps you should add a `mainClass` configuration to jib-maven-plugin -> [Help 1]
```

**`jib-maven-plugin` Configuration**: <!-- Delete this section if not used -->
```xml
<build>
    <plugins>
      <plugin>
          <groupId>com.google.cloud.tools</groupId>
          <artifactId>jib-maven-plugin</artifactId>
          <version>3.0.0</version> <!-- 3.0.0 works, 3.1.0 fails -->
          <configuration>
            <from>
              <image>nginx:alpine</image>
            </from>
            <to>
              <image>a:b</image>
            </to>
            <container>
              <entrypoint>INHERIT</entrypoint>
            </container>
          </configuration>
          <executions>
            <execution>
              <id>docker-image-package</id>
              <phase>package</phase>
              <goals>
                <goal>dockerBuild</goal>
              </goals>
            </execution>
          </executions>
        </plugin>
    </plugins>
</build>
```


**Log output**: <!-- If applicable, provide relevant log output -->

```
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:3.1.0:dockerBuild (docker-image-package) on project javadocs: Main class was not found, perhaps you should add a `mainClass` configuration to jib-maven-plugin -> [Help 1]
```

**Additional Information**: <!-- Any additional information that may be helpful -->


<!-- Thanks for contributing! -->

--- a/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/PluginConfigurationProcessor.java
+++ b/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/PluginConfigurationProcessor.java
@@ -632,9 +632,19 @@ public class PluginConfigurationProcessor {
     }
 
     String classpathString = String.join(":", classpath);
-    String mainClass =
-        MainClassResolver.resolveMainClass(
-            rawConfiguration.getMainClass().orElse(null), projectProperties);
+    String mainClass;
+    try {
+      mainClass =
+          MainClassResolver.resolveMainClass(
+              rawConfiguration.getMainClass().orElse(null), projectProperties);
+    } catch (MainClassInferenceException ex) {
+      if (entrypointDefined) {
+        // We will use the user-given entrypoint, so don't fail.
+        mainClass = "could-not-infer-a-main-class";
+      } else {
+        throw ex;
+      }
+    }
     addJvmArgFilesLayer(
         rawConfiguration, projectProperties, jibContainerBuilder, classpathString, mainClass);
 

--- a/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/PluginConfigurationProcessorTest.java
+++ b/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/PluginConfigurationProcessorTest.java
@@ -496,6 +496,54 @@ public class PluginConfigurationProcessorTest {
     assertThat(jvmArgFiles)
         .comparingElementsUsing(EXTRACTION_PATH_OF)
         .containsExactly("/app/jib-classpath-file", "/app/jib-main-class-file");
+
+    String classpath =
+        new String(
+            Files.readAllBytes(appCacheDirectory.resolve("jib-classpath-file")),
+            StandardCharsets.UTF_8);
+    assertThat(classpath)
+        .isEqualTo("/foo:/app/resources:/app/classes:/app/libs/foo-1.jar:/app/libs/bar-2.jar");
+    String mainClass =
+        new String(
+            Files.readAllBytes(appCacheDirectory.resolve("jib-main-class-file")),
+            StandardCharsets.UTF_8);
+    assertThat(mainClass).isEqualTo("java.lang.Object");
+  }
+
+  @Test
+  public void testClasspathArgumentFile_mainClassInferenceFailureWithCustomEntrypoint()
+      throws NumberFormatException, InvalidImageReferenceException, MainClassInferenceException,
+          InvalidAppRootException, IOException, InvalidWorkingDirectoryException,
+          InvalidPlatformException, InvalidContainerVolumeException,
+          IncompatibleBaseImageJavaVersionException, InvalidContainerizingModeException,
+          InvalidFilesModificationTimeException, InvalidCreationTimeException {
+    when(rawConfiguration.getMainClass()).thenReturn(Optional.of("invalid main class"));
+    when(rawConfiguration.getEntrypoint()).thenReturn(Optional.of(Arrays.asList("bash")));
+
+    ContainerBuildPlan buildPlan = processCommonConfiguration();
+
+    assertThat(buildPlan.getEntrypoint()).containsExactly("bash");
+
+    List<FileEntry> jvmArgFiles = getLayerEntries(buildPlan, "jvm arg files");
+    assertThat(jvmArgFiles)
+        .comparingElementsUsing(SOURCE_FILE_OF)
+        .containsExactly(
+            appCacheDirectory.resolve("jib-classpath-file"),
+            appCacheDirectory.resolve("jib-main-class-file"));
+    assertThat(jvmArgFiles)
+        .comparingElementsUsing(EXTRACTION_PATH_OF)
+        .containsExactly("/app/jib-classpath-file", "/app/jib-main-class-file");
+
+    String classpath =
+        new String(
+            Files.readAllBytes(appCacheDirectory.resolve("jib-classpath-file")),
+            StandardCharsets.UTF_8);
+    assertThat(classpath).isEqualTo("/app/resources:/app/classes:/app/libs/*");
+    String mainClass =
+        new String(
+            Files.readAllBytes(appCacheDirectory.resolve("jib-main-class-file")),
+            StandardCharsets.UTF_8);
+    assertThat(mainClass).isEqualTo("could-not-infer-a-main-class");
   }
 
   @Test

